Add cam:u support and basic documentation.

Steveice10 opened the pull request:
Adds support for almost all of cam:u, as much as I know of anyway.

Crappy basic test that writes very first camera frame to "out.bin" (time constraints):
Binaries: https://www.dropbox.com/s/n7zjso8k3934zcc/CameraTest.zip?dl=1
Source: https://gist.github.com/Steveice10/3f2ddfea716a033b1238

archshift commented on libctru/source/services/cam.c:
```diff
@@ -0,0 +1,737 @@
+
+    if ((ret = svcSendSyncRequest(camHandle)) != 0) return ret;
+    return cmdbuf[1];
+}
```
Newline at EOF

Steveice10 replied:
Gotcha, I'll be making some touch-ups when I get the chance tomorrow. (e.g. Another thing I realized was that I could've just specified bit size for enum fields in structs instead of typing them as u8)

archshift commented:
If you find time, could you at some point add documentation to 3dbrew about cam:u?

Steveice10 commented:
Sure, I can try to do that when I get the time.

archshift commented:
All the structs you added which are serialized for IPC need to be packed with `#pragma pack(1)` so the compiler doesn't mistakenly insert any padding anywhere.

yuriks commented:
@archshift I would change the `bool`s to `u8` to be explicit, but otherwise they're all naturally aligned, so that's not necessary.

Steveice10 commented:
Current issues should be fixed. I added the newline to the end of the file, added "#pragma pack(1)" to the structs, and changed the enum fields from u8 with enum type info in the docs to the enum type with the number of bits specified.

Lectem commented on libctru/include/3ds/services/cam.h:
```diff
@@ -0,0 +1,809 @@
+    float rotationX;
+    /**
+     * @brief Y axis rotation to match the left camera image with the right.
+     */
```
If you only give a brief description use "type member;///< description" syntax. It will be way easier to browse it

yuriks replied:
+1

yuriks commented:
I would not use #pragma pack if it's not necessary (which it isn't here) because it might make the compiler generate more inefficient code, since it changes the struct as a whole to have byte-alignment.

profi200 commented:
Also were got you all the infos from? It"s so detailed and complete that it can't be through reverse engineering or trial and error.

Steveice10 commented:
Documentation issues should be fixed, and pragma pack removed.

The information was gathered and refined over time. It's primarily a mix of RE of functions and their usages, observed behavior, guessing and testing, similarities to Y2R, and identification based on general camera info.

Lectem commented on libctru/include/3ds/services/cam.h:
```diff
@@ -0,0 +1,648 @@
+    CAMU_Contrast contrast : 8;             ///< Camera contrast.
+    CAMU_LensCorrection lensCorrection : 8; ///< Camera lens correction.
+    bool noiseFilterOn;                     ///< Whether to enable the camera's noise filter.
+    int : 8;                                ///< Padding.
```
u8 would be better here

fincs replied:
Those `bool` fields should be `u8` as well (assuming that's the space they take).

Steveice10 replied:
I can go ahead and change the padding to a u8. For the bools, which is more preferred, typing as u8 for size, or typing as bool and using the same syntax as the enum fields to preserve type meaning while ensuring size?

Lectem replied:
Well, thing is, while I like bitfields since it can ensure type safety, the problem with those is that its up to the compiler to (re)arrange them in any order. While GCC usually does what you could expect from it when using byte aligned bitfields, and that we do use those in a few other places, I'd rather use a u8 instead of bool+bitfield.
I also think you should check those structures size and check if the members are indeed in the order you expect them to be if they are to be passed to services commands.

yuriks replied:
@Lectem: Bitfield ordering is implementation-specified and is part of the ABI, so in practice, once you get it right once, it's never going to change.

Lectem replied:
Well yes, but I was thinking more of the bool than the enums, my point was to make sure those fields were in the good order and that the bool did have the good size

Steveice10 commented:
Struct types fixed to match field size, struct order and size double checked against command buffer writes.

profi200 commented:
Did you use that bowling game to get some infos about the cam service?

Steveice10 commented:
That is one source I used, yes. There were a few camera functions there, but not too many.

fincs commented:
Could you recommit this using tabs instead of spaces for indentation?

Steveice10 commented:
Done.
